### PR TITLE
Implement Warm-starting CMA-ES

### PR DESCRIPTION
**Pull request by c-bata**
https://arxiv.org/abs/2012.06932

* [x] Add function to get the mean, the sigma and the covariance matrix.
* [x] Add an argument to set the covariance matrix.
* [x] Reproduce an experiments result of the paper
    * Section 4.1: Performance Depending on Task Similarity: https://gist.github.com/c-bata/896ab75b83b26f560a8fb1eecfb9e232
* [x] Visualizer

### CODE REVIEW

File: cmaes/_warm_start.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import math
+import numpy as np
+
+from typing import Tuple, List
+
+
+def get_sigma_and_mean(
```
Comment (c-bata) [inline cmaes/_warm_start.py:7]:
```suggestion
def get_starting_point(
```

File: cmaes/_warm_start.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    solutions: List[Tuple[np.ndarray, float]],
+    ratio_high_rank_gamma: float = 0.1,
+    normal_prior_alpha: float = 0.1,
+) -> Tuple[np.ndarray, float, np.ndarray]:
+    if len(solutions) == 0:
+        raise ValueError("solutions should contain one or more items.")
+
+    solutions = sorted(solutions, key=lambda t: t[1])
+    gamma_n = int(len(solutions) * ratio_high_rank_gamma)
```
Comment (c-bata) [inline cmaes/_warm_start.py:16]:
```suggestion
    gamma_n = math.ceil(len(solutions) * ratio_high_rank_gamma)
```

**Comment by c-bata** [comment]
https://user-images.githubusercontent.com/5564044/106707840-3eedbe00-6635-11eb-8a0a-bece02173376.mp4


https://user-images.githubusercontent.com/5564044/106707855-4319db80-6635-11eb-9bee-fb593ba6c655.mp4


https://user-images.githubusercontent.com/5564044/106707859-43b27200-6635-11eb-8bfa-9dfe15de5801.mp4


https://user-images.githubusercontent.com/5564044/106707861-444b0880-6635-11eb-8ebf-c94f4bd77eed.mp4

File: cmaes/_warm_start.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+from typing import Tuple, List
+
+
+def get_starting_point(
+    source_solutions: List[Tuple[np.ndarray, float]],
+    gamma: float = 0.1,
+    alpha: float = 0.1,
+) -> Tuple[np.ndarray, float, np.ndarray]:
+    """Estimate a promising distribution of the source task, then
```
Comment (nmasahiro) [inline cmaes/_warm_start.py:12]: [nits] Estimate => Estimates

File: cmaes/_warm_start.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    returns a starting point (mean vector and the covariance matrix
+    of the multivariate gaussian distribution in the CMA-ES).
```
Comment (nmasahiro) [inline cmaes/_warm_start.py:14]: [nits] returns a multivariate gaussian distribution (the mean vector and the covariance matrix) used for initialization of the CMA-ES.

File: cmaes/_warm_start.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+    Args:
+        source_solutions:
+            List of solutions (parameter, value) on a source task.
+
+        gamma:
+            top-gamma solutions are selected from a set of solutions
```
Comment (nmasahiro) [inline cmaes/_warm_start.py:21]: [nits] top-gamma => top-(gamma x 100)%

File: cmaes/_warm_start.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    Returns:
+        The tuple of mean vector, sigma, and covariance matrix.
+    """
+    # Paper: https://arxiv.org/abs/2012.06932
+    assert 0 < gamma <= 1, "gamma should be in (0, 1]"
+
+    if len(source_solutions) == 0:
+        raise ValueError("solutions should contain one or more items.")
+
+    # Select top-gamma solutions
```
Comment (nmasahiro) [inline cmaes/_warm_start.py:36]: top-(gamma x 100)%

**Comment by nmasahiro** [comment]
@c-bata I left some minor comments. Can you check the points?

**Comment by c-bata** [comment]
@nmasahiro Thank you for your review! I applied your suggestions.

**Pull request merged**